### ISSUE TITLE
Investment - Atarashi O

### ISSUE THREAD

**Issue by ladygoogoo** [ladygoogoo/passionprojects#15]
nasdaq_dic = {
    '2020': {'AAPL', 'MSFT', 'AMZN', 'GOOGL', 'META', 'INTC', 'CMCSA', 'CSCO', 'PEP', 'ORCL',
             'ADBE', 'NFLX', 'NVDA', 'PYPL', 'TXN', 'AVGO', 'AMGN', 'GILD', 'BKNG', 'CHTR'},
    '2021': {'AAPL', 'MSFT', 'AMZN', 'GOOGL', 'TSLA', 'META', 'NVDA', 'PYPL', 'ADBE', 'CMCSA',
             'NFLX', 'PEP', 'INTC', 'CSCO', 'TXN', 'AVGO', 'AMGN', 'COST', 'QCOM', 'MDLZ'},
    '2022': {'AAPL', 'MSFT', 'AMZN', 'GOOGL', 'TSLA', 'META', 'NVDA', 'PYPL', 'ADBE', 'NFLX',
             'CMCSA', 'PEP', 'INTC', 'CSCO', 'TXN', 'AVGO', 'AMGN', 'COST', 'QCOM', 'MDLZ'},
    '2023': {'AAPL', 'MSFT', 'AMZN', 'GOOGL', 'TSLA', 'META', 'NVDA', 'ADBE', 'NFLX', 'PYPL',
             'CMCSA', 'PEP', 'INTC', 'CSCO', 'TXN', 'AVGO', 'AMGN', 'COST', 'QCOM', 'MDLZ'},
    '2024': {'AAPL', 'MSFT', 'AMZN', 'GOOGL', 'TSLA', 'META', 'NVDA', 'ADBE', 'NFLX', 'PYPL',
             'CMCSA', 'PEP', 'INTC', 'CSCO', 'TXN', 'AVGO', 'AMGN', 'COST', 'QCOM'}
}

**Comment by ladygoogoo** [comment]
import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime, timedelta
import random

**Comment by ladygoogoo** [comment]
def compute_year(year):

  trades_dic = {}
  for ticker in nasdaq_dic[year]:
    #download data
    ticker_ = yf.Ticker(ticker)
    stock_data = ticker_.history(start=f"{year}-01-01", end=f"{year}-12-31")

    #add ATV
    stock_data['High-Low'] = stock_data['High'] - stock_data['Low']
    stock_data['High-Close'] = abs(stock_data['High'] - stock_data['Close'].shift(1))
    stock_data['Low-Close'] = abs(stock_data['Low'] - stock_data['Close'].shift(1))
    stock_data['TR'] = stock_data[['High-Low', 'High-Close', 'Low-Close']].max(axis=1)
    stock_data['ATV'] = stock_data['TR'].rolling(window=14).mean()

    #chose a random date
    try:
      random_row = np.random.randint(0, len(stock_data))
      random_close = stock_data.iloc[random_row]['Close']
      random_atv = stock_data.iloc[random_row]['ATV']
      max_high = stock_data.iloc[random_row]['High']
      entry_price = random_close
      distance = 2 * random_atv
      stoploss = entry_price - distance
      qty = 2000 / distance
      new_stoploss = stoploss

      stoploss_hit = False
      for i in range(random_row, len(stock_data)):
        high = stock_data.iloc[i]['High']
        low = stock_data.iloc[i]['Low']
        close = stock_data.iloc[i]['Close']
        if max_high < high:
          max_high = high
          new_stoploss = max_high - distance
          continue
        if low < new_stoploss:
          stoploss_hit = True
          break

      #compute all data
      if stoploss_hit == True:
        pnl = (new_stoploss - entry_price) * qty
        trades_dic[ticker] = {'entry_date': stock_data.index[random_row], 'closing_date': stock_data.index[i], 'entry_price': entry_price, 'closing_price': new_stoploss,'quantity': qty, 'pnl': pnl}
      if stoploss_hit == False:
        pnl = (stock_data.iloc[len(stock_data)-1]['Close'] - entry_price) * qty
        trades_dic[ticker] = {'entry_date': stock_data.index[random_row], 'closing_date': stock_data.index[i], 'entry_price': entry_price, 'closing_price': new_stoploss,'quantity': qty, 'pnl': pnl}

    except Exception as e:
      print(f"{e}")
      continue

  return trades_dic

**Comment by ladygoogoo** [comment]
def balance_year(year,trades_dic):

  balance = {}

  date_range = pd.date_range(start=f'{year}-01-01', end=f'{year}-12-31', freq='D')
  date_range_with_tz = date_range.tz_localize('America/New_York')
  date_array = date_range_with_tz.to_list()

  last_pnl = 0
  for date in date_array:
    total_pnl = 0
    for trade in trades_dic.keys():
        data = trades_dic[trade]
        if data['closing_date'] == date:
          total_pnl += data['pnl']
    last_pnl = last_pnl + total_pnl
    balance[date] = last_pnl

  return balance

**Comment by ladygoogoo** [comment]
def plot_year(year, n):

  trade_ = []
  plt.figure(figsize=(12, 6))

  for i in range(n):

    trades_dic = compute_year(year)
    balance = balance_year(year,trades_dic)
    plt.plot(list(balance.keys()), list(balance.values()), alpha=0.2, color='orange')
    trade_.append(list(balance.values()))

  plt.plot(list(balance.keys()), np.mean(trade_, axis = 0), color='black',linewidth=2)
  plt.title(f"Results for {year}")
  plt.grid(axis='y', linestyle='--', alpha=0.7)

  plt.show()

**Comment by ladygoogoo** [comment]
for year in ['2020','2021','2022','2023','2024']:
  plot_year(year,50)
